### PR TITLE
Various Diplo Fixes & Improvements

### PR DESCRIPTION
**Pull request by HeathcliffWarriors**
**CHANGELOG**
Writing this in case I screw up again, so it's clear where the issue is :)

**[Diplomacy AI]**

New Functions
- HasMetValidMinorCiv: Returns true if this player knows a valid minor civ, false otherwise (useful for differentiating AI behavior when City-States are ingame or not, like if you...wanted to delete the Scrivener's Office emissary, for example :P)

- IsPlayerDPWithAnyFriend: Moved the code in GetDPWithAnyFriendScore into its own function (was useful for LuaPlayer changes)

- IsPlayerDPWithAnyEnemy: Moved the code in GetDPWithAnyEnemyScore into its own function (ditto)

- IsPlayerSameIdeology: Moved the code in GetSameLatePoliciesScore into its own function (ditto)

- IsPlayerOpposingIdeology: Moved the code in GetSameLatePoliciesScore into its own function (ditto)

- GetResearchAgreementScore: Exists to add the new opinion modifier to GetMajorCivOpinionWeight

Changes to Promise Logic
- Fixed two bugs with military promise counter.
- AIs are flagged as making a military/expansion promise properly now, so the counter will be visible to humans.
- Expansion and border promise counters changed to 50 turns irrespective of game speed; temporary fix for an asymmetry bug between AIs (one AI would end its expansion promise after 50 turns regardless of gamespeed and the other would still consider it to have made a promise beyond that). Will work on this some more when I can actually see the counters.
- When an AI ends a promise it made for whatever reason (ending a DoF for example), the other AI (or human shadow AI) will flag it as if they didn't make the promise; temporary fix for the same reason, promises could definitely be coded better and I'm working on fixing issues with them.
- Resurrecting a player or having a vassal capitulate to you now resets all promises.

Changes re: Teammates
- Teammates will now always agree not to spread their religion or dig up artifacts
- AI should no longer try to bully/conquer City-States that their teammates have allied or protected
- Removed penalties for breaking a coop war promise with a teammate
- Removed several opinion penalties for teammates so they show up as white/don't show up in the diplo mood UI

Changes re: Coop Wars
- When coop war state is set to COOP_WAR_STATE_ACCEPTED, it removes -1 from GetNumTimesCoopWarDenied (if above 0); this makes it possible to recover from this penalty by going to war with the AI later on, whereas previously there was no recovery mechanic at all.
- SetNumTimesCoopWarDenied to 0 on resurrection or capitulation.
- When either the human or the AI warns the intended target of a coop war, their teammates and DPs are now warned as well. The target and their allies will gain a diplo penalty towards the asker, and a diplo bonus towards the person who warned them (for allies, based on proximity to the target player).

GetBestApproachTowardsMajorCiv Changes
- Moved "Coop War Denied?" check to just after the DoF check, because of the next change.
- Resurrecting a player now has the same effect as a DoF (+FRIENDLY, WAR/HOSTILE = 0 early on in the function), and also zeroes out DECEPTIVE weight for declining coop war requests; small buff to resurrecting a player, and deception towards a liberator over this is kind of dumb
- Added MinorCivDisputeLevel approach check (only if the AI has met a valid City-State); AIs going for diplo victory would want to check for this, and it's the only dispute level check that's missing.
- Fixed a potential bug with master zero-out hostility weight towards vassals; moved it to an Else If check.
- Added FRIENDLY bias x1 for being a voluntary vassal. Rationale: A voluntary protector should get some extra friendliness, matches up with the reduction to denouncing weight for voluntary vassals, and also a small buff to resurrecting a player in the Medieval Era or later (they are likely to be VERY weak and thus friendliness towards their liberator/protector is intelligent)
- Moved "On the same team?" check to the end of the function, fixing a potential issue where teammates could have a different approach surpass FRIENDLY if Opinion was high enough

GetMajorCivOpinionWeight
- Added -5 opinion if you have a research agreement with the AI. Rationale: It already affects approach, it's good to indicate to the player that it's affecting AI behavior, and it makes sense to have the AI be a slight bit more friendly towards people who are helping them gain in Science.

DoEstimateOtherPlayerOpinions
- Added research agreement modifier

Help Requests
- Now only resets the help request timer if the AI agreed to the request. Unlike demands, there isn't a reason why the timer should be set otherwise, since the AI is either willing to give you something or it isn't; bullying has no effect. Will also fix the issue where they say "It's been too soon since our last show of generosity." when they haven't given you anything.

DoSomeoneDeclaredWarOnMe
- Now sets approach guess and counter instantly rather than waiting a turn

DoWeMadePeaceWithSomeone
- SetWantsSneakAttack = false

DoContactPlayer
- AI should no longer send certain insulting statements to players who have resurrected them

Other
- Resurrected players will now always accept requests to not settle/spy/convert/dig up artifacts from their liberator.
- Fixed bug where capitulated vassals didn't reset no settle/spy request penalties properly due to CBP changes
- Cleared warmonger penalty for capitulated vassals; this was originally intended by Putmalk, and as capitulation essentially puts a player out of the game, it's unfun/dumb for them to keep holding a grudge over past warmongering
- Vassals no longer want a sneak attack towards their master after vassalage is made
- Resets all promises upon vassal capitulation
- Stopped ending a DoF with an AI planning war towards you from disrupting their war plans (uses WAR_FACE_GUARDED instead of setting approach to GUARDED), and AI also updates its opinion instantly
- Likewise for denouncing a friend; cleaned up DoDenouncePlayer logic a bit

[Other Stuff]

Text
- Clarified military promise TT
- Added counter display for AI military/expansion promises to LuaPlayer
- Incorporated community typo fixes per this thread (https://forums.civfanatics.com/threads/text-updates.647903/)
- Reorganized CoreText_en_US.sql file so everything is in a more sensible order
- Started correcting spelling/grammar errors in leader dialog, some of these fixes have been included
- Fixed issue with certain text fixes not being applied ingame
- Fixed missing periods in game option descriptions for events
- Added text key for GetResearchAgreementScore, see below
- Modified coop war intrigue sharing text keys in accordance with changes above
- Copied the typo fixes from the CBP version of C4DF leader dialog to the other two files, just because
- Misc. comment typo fixes & improvements

DealAI
- Empty requests will no longer trigger the help request timer (I set it to DEMAND_RESPONSE_GIFT_REFUSE_TOO_MUCH, which isn't an ideal fix but works for now)

DealClasses
- Set recent assist change for brokered war/peace to 300, since higher values make no difference

CvPlayer
- Fixed bug where no settle/no spy request penalties weren't being cleared properly due to CBP changes
- Resurrecting a player now resets all promises

LuaPlayer
- Should now display AI promises for military/expansion as well as human ones
- Added Research Agreement modifier
- Certain modifiers that are now set to 0 for teammates are still displayed (if they exist) to give the player useful information: making a DoF/DP with a teammate's enemy, denouncing a teammate's friend, or having a different Ideology. The text is now shown in white rather than red.

Misc
- Changed the order of a few sections of code without changing its functioning, just for easier readability

### CODE REVIEW

File: Community Balance Patch/Balance Changes/Text/en_US/BuildingText.sql [hunk 1]
```diff
@@ -92,7 +92,7 @@ SET Text = '+20 XP for all Units. -1 [ICON_HAPPINESS_3] Unhappiness from [ICON_F
 WHERE Tag = 'TXT_KEY_BUILDING_ARMORY_HELP' AND EXISTS (SELECT * FROM COMMUNITY WHERE Type='COMMUNITY_CORE_BALANCE_BUILDINGS' AND Value= 1 );
 
 UPDATE Language_en_US
-SET Text = '-1 [ICON_HAPPINESS_3] Unhappiness from [ICON_FOOD]/[ICON_PRODUCTION] Distress. +25 XP for all Units.[NEWLINE][NEWLINE]City must have an Armory. Increases the Military Unit Supply Cap by 1.'
+SET Text = '-1 [ICON_HAPPINESS_3] Unhappiness from [ICON_FOOD]/[ICON_PRODUCTION] Distress. +15% [ICON_PRODUCTION] towards Land Units, and +25 XP for all Units.[NEWLINE][NEWLINE]City must have an Armory. Increases the Military Unit Supply Cap by 1.'
```
Comment (Aqinu) [inline Community Balance Patch/Balance Changes/Text/en_US/BuildingText.sql]: Forgot to add after [ICON_PRODUCTION] the "Production" word

File: Community Balance Patch/Balance Changes/Text/en_US/PolicyText.sql [hunk 2]
```diff
@@ -961,11 +961,11 @@ WHERE Tag = 'TXT_KEY_POLICY_FREERELIGION_TEXT' AND EXISTS (SELECT * FROM COMMUNI
 
 -- Monastery
 UPDATE Language_en_US
-SET Text = '+3 [ICON_FOOD] Food, +3 [ICON_RESEARCH] Science, +1 [ICON_PEACE] Faith.'
+SET Text = 'Religious building purchased with [ICON_PEACE] Faith. Increases City [ICON_FOOD] Food, [ICON_RESEARCH] Science and [ICON_PEACE] Faith output. Requires the adoption of the Fealty Policy Branch.'
```
Comment (Aqinu) [inline Community Balance Patch/Balance Changes/Text/en_US/PolicyText.sql]: Why did you remove the quantity of each bonus?

Comment (HeathcliffWarriors) [reply Community Balance Patch/Balance Changes/Text/en_US/PolicyText.sql]: https://forums.civfanatics.com/threads/text-updates.647903/#post-15498779

File: Community Balance Patch/Balance Changes/Text/en_US/BuildingText.sql (hunk 1 above)
Comment (HeathcliffWarriors) [reply Community Balance Patch/Balance Changes/Text/en_US/BuildingText.sql]: Fixed, thank you for pointing it out :)

**Comment by LoneGazebo** [comment]
Unrest is fine by me.

**Comment by LoneGazebo** [comment]
@HeathcliffWarriors just FYI I had to do quite a bit of cleanup on this, as you're calling functions that don't exist in some situations. Are you doing your edits in VS? If not, you should be, it'll warn you if you're calling a nonexistent function.

G

**Comment by HeathcliffWarriors** [comment]
@LoneGazebo As you seem to have guessed, I haven't been - I looked for information on how to get Visual Studio and the VP DLL to work on Windows 10 for editing purposes, but I haven't been able to find any good information yet and I've been left confused. I recall there used to be some kind of tutorial on how to contribute to VP that doesn't seem to be present anymore, so I've been emulating (with modifications) the existing code and what I've seen of others' contributions to figure out what to do.

I see that's caused an issue, sorry for causing you trouble.

Do you have any information on the required steps to make edits in Visual Studio? I certainly don't mind doing so, I just don't know how yet :)

**Comment by LoneGazebo** [comment]
https://forums.civfanatics.com/threads/how-to-build-and-contribute-to-the-community-patch-dll.528319/

**Comment by LoneGazebo** [comment]
This is still largely accurate.
